driver-core/src/test/unit/com/mongodb/internal/connection/ConnectionPoolTest.java

```diff
@@ -0,0 +1,333 @@
+                    Exception exceptionFromFuture = futureMap.remove(target).get(5, TimeUnit.SECONDS);
+                    if (exceptionFromFuture != null) {
+                        throw exceptionFromFuture;
+                    }
+                } else if (name.equals("wait")) {
+                    Thread.sleep(operation.getNumber("ms").intValue());
+                } else if (name.equals("clear")) {
+                    pool.invalidate();
+                } else if (name.equals("close")) {
+                    pool.close();
+                } else {
+                    Callable<Exception> callable = createCallable(operation);
+                    if (operation.containsKey("thread")) {
+                        String threadTarget = operation.getString("thread").getValue();
+                        ExecutorService executorService = executorServiceMap.get(threadTarget);
+                        futureMap.put(threadTarget, executorService.submit(callable));
+                    } else {
+                        callable.call();
+                    }
+                }
+            }
+        } catch (Exception e) {
+            if (!definition.containsKey("error")) {
+                throw e;
+            }
+            BsonDocument errorDocument = definition.getDocument("error");
+            String exceptionType = errorDocument.getString("type").getValue();
+            if (exceptionType.equals("PoolClosedError")) {
+                assertEquals(IllegalStateException.class, e.getClass());
+            } else if (exceptionType.equals("WaitQueueTimeoutError")) {
+                assertEquals(MongoTimeoutException.class, e.getClass());
+            } else {
+                fail("Unexpected exception type " + exceptionType);
+            }
+        }
+
+        if (definition.containsKey("events")) {
+            Iterator<Object> actualEventsIterator = getNonIgnoredActualEvents().iterator();
+            BsonArray expectedEvents = definition.getArray("events");
+            for (BsonValue cur : expectedEvents) {
+                BsonDocument expectedEvent = cur.asDocument();
+                String type = expectedEvent.getString("type").getValue();
+                if (type.equals("ConnectionPoolCreated")) {
+                    ConnectionPoolOpenedEvent actualEvent = getNextEvent(actualEventsIterator, ConnectionPoolOpenedEvent.class);
+                    assertEquals(serverAddress, actualEvent.getServerId().getAddress());
+                    assertEquals(settings, actualEvent.getSettings());
+                } else if (type.equals("ConnectionPoolClosed")) {
+                    ConnectionPoolClosedEvent actualEvent = getNextEvent(actualEventsIterator, ConnectionPoolClosedEvent.class);
+                    assertEquals(serverAddress, actualEvent.getServerId().getAddress());
+                } else if (type.equals("ConnectionPoolCleared")) {
+                    // TODO
```

Comment:
The TODOs are for 4.0, when we can implement the remaining events

Reply:
Suggest being explicit about that eg: `// TODO in 4.0`

Reply:
Done